Fix ordering of context.check_hostname and context.verify_mode

coltonbh opened the pull request:
Closes #356 

Please let me know if there are other security concerns to address here. This appears to be a bug since the code suggests that `context.verity_mode = ssl.CERT_NONE` and `context.check_hostname = False` should be a valid state.

coltonbh commented:
Also, this is a small contribution but if appropriate would be cool to add myself to the AUTHORS list. Been using `celery` for a long time--would be sweet to be an official contributor to part of it :)

auvipy commented:
@moisesguimaraes

auvipy reviewed (commented):
thanks for handling this. since changing the order fixes the issue, don't we need adjustment in tests as well? also a changelog entry is very much appreciated

coltonbh commented:
@auvipy Great! Happy to add a changelog entry. Re: tests--any pointers on exactly what you'd like done here? If you can point me to some tests that would exercise this specific piece of code I'd be happy to write a test or two. I'm not clear on the full context in which this class gets exercised.

auvipy commented:
related tests are located here https://github.com/celery/py-amqp/blob/master/t/unit/test_transport.py

coltonbh commented:
Great. I'll have a look and see what I can do. Thanks!

moisesguimaraes commented:
He y'all,

Thanks for the ping @auvipy.

I have something here to consider:

```
In [1]: import ssl

In [2]: context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)

In [3]: context.check_hostname
Out[3]: False

In [4]: context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
   ...: 

In [5]: context.check_hostname
Out[5]: True
```

We should probably cover both scenarios with tests here.

moisesguimaraes commented:
A few more tests

```
In [29]: context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)

In [30]: context.verify_mode
Out[30]: <VerifyMode.CERT_NONE: 0>

In [31]: context.check_hostname = True

In [32]: context.verify_mode
Out[32]: <VerifyMode.CERT_REQUIRED: 2>

In [33]: context.verify_mode = ssl.CERT_NONE
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
```

I'm happy with the current changes, @coltonbh++.

matusvalo commented:
I am happy with changes. The only missing point are tests.

coltonbh commented:
@moisesguimaraes @auvipy @matusvalo Changelog entry added and tests added. I'm not quite sure what to assert on the tests since the previous tests mistakenly passed due to the entire `context` object being mocked. Here my goal is to leave as much of the `context` object's default behavior intact and assure than the code can execute successfully. Do you have further suggestions on how to improve these tests? Thanks!

matusvalo commented:
@coltonbh the one possible solution how test properly this is to test it via integration tests:

https://github.com/celery/py-amqp/blob/master/t/integration/test_rmq.py

coltonbh commented:
@matusvalo I'm trying to test that the modifications on the `context` object performed in `_wrap_socket_sni` do not raise exceptions (as previously they did). My tests do fail under the old code (at least the `ssl.PROTOCOL_TLS_CLIENT` version does since it sets `check_hostname=True`). After my question above I added the `mock_wrap_socket.assert_called_once()` check which assures that the code block executed without error and called the `wrap_socket()` method. Do we want to add integration tests? I'm not sure what it would add to actually connect to a RabbitMQ instance since that wasn't the original problem. Please let me know if these feel urgent--I do believe the current tests--with my final assert statement--cover the bug :)

moisesguimaraes commented:
Let me have a quick look

moisesguimaraes commented:
I would move your tests to the end of the function and mock both `load_default_certs` and `wrap_socket` as these would be the expensive calls. So first we test the code path, then we test if `ssl.SSLContext` behaves accordingly.

I would favor tweaking `server_side` param instead of `ssl_version` to set the side of the connection.

Let's remember that checking the hostname with SNI is a client side feature only. There is no way for servers to validate a client's hostname as they don't have one. Focus the test on exercising the possibility to set cert_reqs to both `ssl.CERT_NONE` and `ssl.CERT_REQUIRED` in both scenarios (client side and server side), in other words the test should cover the 4 scenarios of their combinations.

Having the ssl.SSLContext constructor not mocked should fail with the old code and work with the new one. I don't see the need for integration tests at this point.

Looking again at your code, you're almost there. I strongly consider mocking `load_default_certs` as it is an expensive call when loading all system certs.

Thanks again for your contribution @coltonbh o/

thedrow commented on Changelog:
```diff
@@ -4,6 +4,16 @@ Changes
 py-amqp is fork of amqplib used by Kombu containing additional features and improvements.
 The previous amqplib changelog is here:
 http://code.google.com/p/py-amqplib/source/browse/CHANGES
+.. _version-5.0.6:
```
```suggestion

.. _version-5.0.6:
```

coltonbh replied:
Fixed!

thedrow commented on amqp/transport.py:
```diff
@@ -535,6 +533,9 @@ def _wrap_socket_sni(self, sock, keyfile=None, certfile=None,
         except AttributeError:
             pass  # ask forgiveness not permission
 
+        if cert_reqs is not None:
```
If you don't mind, let's document inline what each condition is meant to do and why their order is this way and not another.
Otherwise this issue may regress.

thedrow commented on t/unit/test_transport.py:
```diff
@@ -700,7 +700,22 @@ def test_wrap_socket_ciphers(self):
+        # cert_reqs == ssl.CERT_NONE
+        with patch('ssl.SSLContext.wrap_socket') as mock_wrap_socket:
+            sock = Mock()
+            self.t._wrap_socket_sni(
```
Usually, we don't test the internals of a class.
We exercise the public API and check the outcomes.
This is because this test may pass but another code path in the public method will still make the desired functionality unobtainable.

thedrow reviewed (changes requested):
I agree that integration tests will help us cover this case.
Overall, this is good work. See my comments.

coltonbh commented:
ok @moisesguimaraes and @thedrow your suggestions have been incorporated. Specifically:

- Added the additional tests suggested by @moisesguimaraes so the the full 2x2 matrix of possibilities is covered
- Mocked out `load_default_certs` and asserted correct behavior for this call
- Moved tests to the end of the `test_wrap_socket_sni_cert_reqs` code block
- Added comments inline to clarify why the code must execute in a particular path.

Let me know if anything else is needed!

coltonbh commented:
🎉🥳
